packages/utils/src/cache/distributed-cache.ts

```diff
@@ -0,0 +1,300 @@
+   * Returns cache data value based on provided key
+   * @param key Cache key
+   * @returns Promise string data or `null` if nothing is cached
+   */
+  public async get (key: string): Promise<string | null> {
+    const iconName: string = key.split('/').pop() || '';
+    const item = await this.storage.get(key) as CacheItem;
+
+    if (item && item.expires > Date.now()) {
+      Logger.log(`${window.name} %c Found Cache %c ${iconName} ${Date.now()}`, 'background: lightgreen; color: white', '');
+      return item.value;
+    }
+
+    // Check key is already started a request by using mutex to prevent race condition in multi-threads
+    const newRequest = await mutex.runExclusive(() => {
+      if (!this.hasRequest(key)) {
+        this.addRequest(key);
+        Logger.log(`${window.name} %c Request %c ${iconName} ${Date.now()}`, 'background: blue; color: white', '');
+        return true;
+      }
+      return false;
+    });
+
+    if (newRequest === false) {
+      Logger.log(`${window.name} %c Wait %c ${iconName} ${Date.now().toString()}`, 'background: orange; color: white', '');
+      // Add to waiting list by create promise waiting for resolve
+      return new Promise<string | null>(resolve => {
+        this.wait(key, resolve);
+      });
+    }
+    else {
+      return null;
+    }
+  }
+
+  /**
+   * Caches a value to active cache without writing to storage
+   * @param key Cache key
+   * @param value Data to store in cache
+   * @param [expires=432000] Cache expiry in seconds. Defaults to 5 days.
+   * @returns {void}
+   */
+  protected syncActive (key: string, value: string, expires = 432000): void {
+    const modified = Date.now();
+    const data = {
+      value: value,
+      modified: Date.now(),
+      expires: modified + expires * 1000
+    };
+
+    this.storage.setActive(key, data);
+  }
+
+  /**
+   * Add the started request to the state that is used in checking a duplicated request across other messengers
+   * @param key item key
+   * @return {void}
+   */
+  protected addRequest (key: string): void {
+    const requests = this.requestsState;
+    if (!requests[key]) {
+      requests[key] = 'true';
+      this.requests[key] = 'true';
+      this.requestsState = requests;
+    }
+  }
+
+  /**
+   * Check key is already started request across messengers
+   * @param key resource name
+   * @returns true if resource has started request
+   */
+  protected hasRequest (key: string): boolean {
+    return key in this.requests ? true : key in this.requestsState;
+  }
+
+  /**
+   * Add an item to the waiting list
+   * @param key item key
+   * @param value callback function
+   * @returns {void}
+   */
+  protected wait (key: string, value: CallableFunction): void {
+    this.waiting.set(key, value);
+  }
+
+
+  /**
+   * Clean up all temporary states
+   * @returns {void}
+   */
+  private clean (): void {
+    localStorage.removeItem(this.storageNames.requests);
+    this.requests = {};
+    this.waiting.clear();
```

Comment:
I would keep either `clean` or `clear` not both.

Reply:
The name should be `clean` because the action of this method is clean up the unused temporary data not clear all data, and the other reason is we cannot use `clear` because the method `clear` already exists in class which is used to clear all cache items.